packages/client/src/links/wsLink/wsClient/wsClient.ts

```diff
@@ -0,0 +1,442 @@
+              observer.error(TRPCClientError.from(transformed.error));
+              return;
+            }
+
+            observer.next({
+              result: transformed.result,
+            });
+          },
+        },
+      );
+
+      return () => {
+        abort();
+
+        if (type === 'subscription' && this.activeConnection.isOpen()) {
+          this.send({
+            id,
+            method: 'subscription.stop',
+          });
+        }
+
+        signal?.removeEventListener('abort', abort);
+      };
+    });
+  }
+
+  public get connection() {
+    return backwardCompatibility(this.activeConnection);
+  }
+
+  /**
+   * Manages the reconnection process for the WebSocket using retry logic.
+   * Ensures that only one reconnection attempt is active at a time by tracking the current
+   * reconnection state in the `reconnecting` promise.
+   */
+  private reconnecting: Promise<void> | null = null;
+  private reconnect(closedError: TRPCWebSocketClosedError) {
+    this.connectionState.next({
+      type: 'state',
+      state: 'connecting',
+      error: TRPCClientError.from(closedError),
+    });
+    if (this.reconnecting) return;
+
+    const tryReconnect = async (attemptIndex: number) => {
+      try {
+        await sleep(this.reconnectRetryDelay(attemptIndex));
+        if (this.allowReconnect) {
+          await this.activeConnection.close();
+          await this.activeConnection.open();
+        }
+        this.reconnecting = null;
+      } catch {
+        await tryReconnect(attemptIndex + 1);
+      }
+    };
+
+    this.reconnecting = tryReconnect(0);
+  }
+
+  private setupWebSocketListeners(ws: WebSocket) {
+    const handleCloseOrError = (cause: unknown) => {
+      const reqs = this.requestManager.getPendingRequests();
+      for (const { message, callbacks } of reqs) {
+        if (message.method === 'subscription') continue;
+
+        callbacks.error(
+          TRPCClientError.from(
+            cause ??
+              new TRPCWebSocketClosedError({
+                message: 'WebSocket closed',
+                cause,
+              }),
+          ),
+        );
+        this.requestManager.delete(message.id);
+      }
+    };
+
+    ws.addEventListener('open', async () => {
+      if (this.lazyMode) {
+        this.inactivityTimeout.start();
+      }
+
+      if (this.connectionParams) {
+        try {
+          ws.send(await buildConnectionMessage(this.connectionParams));
+        } catch (error) {
+          ws.close(3000);
+          handleCloseOrError(error);
+          return;
+        }
+      }
+
+      this.callbacks.onOpen?.();
+
+      this.connectionState.next({
+        type: 'state',
+        state: 'pending',
+        error: null,
+      });
+
+      const messages = this.requestManager
+        .getPendingRequests()
+        .map(({ message }) => message);
+      if (messages.length) {
+        ws.send(JSON.stringify(messages.length === 1 ? messages[0] : messages));
```

Comment:
I don't know why we don't always send as batch (but I think this code is unchanged from before)
```suggestion
        ws.send(JSON.stringify(messages));
```